core/index/index_writer.cpp

```diff
@@ -2238,20 +2217,29 @@ IndexWriter::PendingContext IndexWriter::PrepareFlush(
       if (segment_ctx.segment.has_replace_) {
         segment_ctx.MaskUnusedReplace(committed_tick_, tick);
       }
-      auto document_mask = segment_ctx.MakeDocumentMask(tick);
-      if (segment_ctx.flushed.meta.docs_count == document_mask.size()) {
+      DocumentMask document_mask;
+      const auto state = segment_ctx.MakeDocumentMask(tick, document_mask);
+      if (state == FlushedSegmentContext::State::Skip) {
         continue;
       }
       // TODO(MBkkt) we can avoid copy if it is last usage of FlushedSegment
       IndexSegment new_segment = segment_ctx.flushed;
-
-      if (!document_mask.empty()) {  // Write non-empty document mask
+      bool need_update = segment_ctx.flushed.need_update &&
+                         state == FlushedSegmentContext::State::Last;
+      if (need_update) {
+        new_segment.meta.version += 4;  // TODO(MBkkt) should be 2
+      }
+      // TODO(MBkkt) unnecessary write mask in case if mask was same
+      if (!document_mask.empty()) {
         WriteDocumentMask(dir, new_segment.meta, document_mask);
+        need_update = true;
+      }
+      if (need_update) {
         // Write updated segment metadata
         index_utils::FlushIndexSegment(dir, new_segment);
-        // Refresh reader, TODO(MBkkt) move first two args?
-        segment_ctx.reader = std::make_shared<const SegmentReaderImpl>(
-          *segment_ctx.reader, new_segment.meta, std::move(document_mask));
+        // Refresh reader
+        segment_ctx.reader->Update(dir, new_segment.meta,
+                                   std::move(document_mask));
```

Comment:
If segment was split, we need to move version forward and write new segment metadata